ruler/scheduler.go

```diff
@@ -0,0 +1,222 @@
+	}, []string{"operation", "status_code"})
+)
+
+func init() {
+	prometheus.MustRegister(queueLength)
+	prometheus.MustRegister(configsRequestDuration)
+	prometheus.MustRegister(totalConfigs)
+}
+
+type workItem struct {
+	userID    string
+	rules     []rules.Rule
+	scheduled time.Time
+}
+
+// Key implements ScheduledItem
+func (w workItem) Key() string {
+	return w.userID
+}
+
+// Scheduled implements ScheduledItem
+func (w workItem) Scheduled() time.Time {
+	return w.scheduled
+}
+
+// Defer returns a copy of this work item, rescheduled to a later time.
+func (w workItem) Defer(interval time.Duration) workItem {
+	return workItem{w.userID, w.rules, w.scheduled.Add(interval)}
+}
+
+type scheduler struct {
+	configsAPI         configsAPI // XXX: Maybe make this an interface ConfigSource or similar.
+	evaluationInterval time.Duration
+	q                  *util.SchedulingQueue
+
+	// All the configurations that we have. Only used for instrumentation.
+	cfgs map[string]cortexConfig
+
+	pollInterval time.Duration
+
+	latestConfig configID
+	latestMutex  sync.RWMutex
+
+	done       chan struct{}
+	terminated chan struct{}
+}
+
+// newScheduler makes a new scheduler.
+func newScheduler(configsAPI configsAPI, evaluationInterval, pollInterval time.Duration) scheduler {
+	return scheduler{
+		configsAPI:         configsAPI,
+		evaluationInterval: evaluationInterval,
+		pollInterval:       pollInterval,
+		q:                  util.NewSchedulingQueue(clockwork.NewRealClock()),
+		cfgs:               map[string]cortexConfig{},
+	}
+}
+
+// Run polls the source of configurations for changes.
+func (s *scheduler) Run() {
+	log.Debugf("Scheduler started")
+	defer close(s.terminated)
+	// Load initial set of all configurations before polling for new ones.
+	s.addNewConfigs(time.Now(), s.loadAllConfigs())
+	ticker := time.NewTicker(s.pollInterval)
+	for {
+		select {
+		case now := <-ticker.C:
+			err := s.updateConfigs(now)
+			if err != nil {
+				log.Warnf("Scheduler: error updating configs: %v", err)
+			}
+		case <-s.done:
+			ticker.Stop()
+		}
+	}
+}
+
+func (s *scheduler) Stop() {
+	close(s.done)
+	s.q.Close()
+	<-s.terminated
+	log.Debugf("Scheduler stopped")
+}
+
+// Load the full set of configurations from the server, retrying with backoff
+// until we can get them.
+func (s *scheduler) loadAllConfigs() map[string]cortexConfigView {
+	backoff := minBackoff
+	for {
+		cfgs, err := s.poll()
+		if err == nil {
+			log.Debugf("Scheduler: found %d configurations in initial load", len(cfgs))
+			return cfgs
+		}
+		log.Warnf("Scheduler: error fetching all configurations, backing off: %v", err)
+		time.Sleep(backoff)
+		backoff *= 2
+		if backoff > maxBackoff {
+			backoff = maxBackoff
+		}
+	}
+}
+
+func (s *scheduler) updateConfigs(now time.Time) error {
+	cfgs, err := s.poll()
+	if err != nil {
+		return err
+	}
+	s.addNewConfigs(now, cfgs)
+	return nil
+}
+
+func (s *scheduler) poll() (map[string]cortexConfigView, error) {
+	s.latestMutex.RLock()
```

Comment:
Is this lock needed? `poll()` is the only place that accesses `s.latestConfig` and it never gets called in parallel.